libyul/backends/evm/OptimizedEVMCodeTransform.cpp

```diff
@@ -0,0 +1,527 @@
+					": too deep in the stack by " + to_string(deficit) + " slots in " + stackToString(m_stack);
+				m_stackErrors.emplace_back(StackTooDeepError(
+					m_currentFunctionInfo ? m_currentFunctionInfo->function.name : YulString{},
+					varNameDeep.empty() ? varNameTop : varNameDeep,
+					deficit,
+					msg
+				));
+				m_assembly.markAsInvalid();
+			}
+		},
+		// Push or dup callback.
+		[&](StackSlot const& _slot)
+		{
+			yulAssert(static_cast<int>(m_stack.size()) == m_assembly.stackHeight(), "");
+
+			// Dup the slot, if already on stack and reachable.
+			if (auto depth = util::findOffset(m_stack | ranges::views::reverse, _slot))
+			{
+				if (*depth < 16)
+				{
+					m_assembly.appendInstruction(evmasm::dupInstruction(static_cast<unsigned>(*depth + 1)));
+					return;
+				}
+				else if (!canBeFreelyGenerated(_slot))
+				{
+					int deficit = static_cast<int>(*depth - 15);
+					YulString varName = slotVariableName(_slot);
+					string msg = (varName.empty() ? "Slot " + stackSlotToString(_slot) : "Variable " + varName.str())
+						+ " is " + to_string(*depth - 15) + " too deep in the stack " + stackToString(m_stack);
+					m_stackErrors.emplace_back(StackTooDeepError(
+						m_currentFunctionInfo ? m_currentFunctionInfo->function.name : YulString{},
+						varName,
+						deficit,
+						msg
+					));
+					m_assembly.markAsInvalid();
+					m_assembly.appendConstant(u256(0xDEADBEEF));
+					return;
+				}
+				// else: the slot is too deep in stack, but can be freely generated, we fall through to push it again.
+			}
+
+			// The slot can be freely generated or is an unassigned return variable. Push it.
+			std::visit(util::GenericVisitor{
+				[&](LiteralSlot const& _literal)
+				{
+					m_assembly.setSourceLocation(locationOf(_literal));
+					m_assembly.appendConstant(_literal.value);
+				},
+				[&](FunctionReturnLabelSlot const&)
+				{
+					yulAssert(false, "Cannot produce function return label.");
+				},
+				[&](FunctionCallReturnLabelSlot const& _returnLabel)
+				{
+					if (!m_returnLabels.count(&_returnLabel.call.get()))
+						m_returnLabels[&_returnLabel.call.get()] = m_assembly.newLabelId();
+					m_assembly.setSourceLocation(locationOf(_returnLabel.call.get()));
+					m_assembly.appendLabelReference(m_returnLabels.at(&_returnLabel.call.get()));
+				},
+				[&](VariableSlot const& _variable)
+				{
+					if (m_currentFunctionInfo)
+						if (util::contains(m_currentFunctionInfo->returnVariables, _variable))
+						{
+							m_assembly.setSourceLocation(locationOf(_variable));
+							m_assembly.appendConstant(0);
+							return;
+						}
+					yulAssert(false, "Variable not found on stack.");
+				},
+				[&](TemporarySlot const&)
+				{
+					yulAssert(false, "Function call result requested, but not found on stack.");
+				},
+				[&](JunkSlot const&)
+				{
+					// Note: this will always be popped, so we can push anything.
+					// TODO: discuss if PC is in fact a good choice here.
```

Comment:
I'm not convinced :)
PC is evil, it could actually break some static analysis tools.
I think CHAINID would be a better deterministic opcode, but of course this cannot be used in all evm versions.
DUP1 would be best IMO.
CODESIZE might be another option.

Reply:
I'm not sure ``DUP1`` is less confusing actually. For something like ``CODESIZE`` that otherwise only rarely occurs you can learn and then have in mind that it's just stack alignment. For ``DUP1`` that's much more hidden.
And it's slightly more expensive...
But all is fine and the cost doesn't differ by much: https://gist.github.com/ekpyron/391401bf29ef1bc1a2a13c11d5b75d7f

Reply:
I brought up the idea a few days ago that since this uses `m_assembly`, we could just introduce `pushJunk` and at least for someone inspecting the assembly stream, it would be clear that this is a special push.

The assembler then could translate that into `DUP1` or `PUSH 1` depending on the stack height.

Reply:
I'm not sure... I generally don't like having an assembly item for it - that's strain on tooling and seems like overkill.
I also said I wouldn't want assembling to depend on evm version (you did suggest to translate the assembly item to codesize, chainid, resp. push0 once maybe available depending on version) - but that won't apply for just emitting DUP1 and PUSH 1 of course - but it'd still be an annoying assembly item, since it wouldn't have fixed code size :-).